flink/src/main/java/org/apache/iceberg/flink/data/vectorized/VectorizedFlinkOrcReaders.java

```
@@ -0,0 +1,278 @@
+    private long batchOffsetInFile;
+
+    LongOrcColumnVector(long batchOffsetInFile) {
+      this.batchOffsetInFile = batchOffsetInFile;
+    }
+
+    @Override
+    public boolean isNullAt(int i) {
+      return false;
+    }
+
+    @Override
+    public long getLong(int i) {
+      return batchOffsetInFile + i;
+    }
+  }
+
+  private static class StructConverter implements Converter {
+    private final Types.StructType structType;
+    private final List<Converter> fieldConverters;
+    private final Map<Integer, ?> idToConstant;
+
+    private StructConverter(Types.StructType structType, List<Converter> fieldConverters,
+                            Map<Integer, ?> idToConstant) {
+      this.structType = structType;
+      this.fieldConverters = fieldConverters;
+      this.idToConstant = idToConstant;
+    }
+
+    @Override
+    public ColumnVector convert(org.apache.orc.storage.ql.exec.vector.ColumnVector vector, int batchSize,
+                                long batchOffsetInFile) {
+      StructColumnVector structVector = (StructColumnVector) vector;
+      List<Types.NestedField> fields = structType.fields();
+      ColumnVector[] fieldVectors = new ColumnVector[fields.size()];
+      for (int pos = 0; pos < fields.size(); pos++) {
+        Types.NestedField field = fields.get(pos);
+        if (idToConstant.containsKey(field.fieldId())) {
+          fieldVectors[pos] = getConstantVectors(field.type(), idToConstant.get(field.fieldId()));
+        } else if (field.equals(MetadataColumns.ROW_POSITION)) {
+          fieldVectors[pos] = new LongOrcColumnVector(batchOffsetInFile);
+        } else {
+          fieldVectors[pos] = fieldConverters.get(pos)
+              .convert(structVector.fields[pos], batchSize, batchOffsetInFile);
+        }
+      }
+
+      return new FlinkRowColumnVector(fieldVectors, structVector);
+    }
+
+    private ColumnVector getConstantVectors(Type type, Object constant) {
+      if (type.isPrimitiveType()) {
```

We don't need this `type.isPrimitiveType()` any more because we've checked each data type in the following switch-case sentence.  We can throw an exception when we encounter primitive types that cannot be handled.